pkg/stub/handler.go

```diff
@@ -113,60 +116,67 @@ type Handler struct {
-		} else {
-			for _, arch := range srcfg.Spec.Architectures {
-				switch arch {
-				case v1alpha1.X86Architecture:
-					cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
-				case v1alpha1.PPCArchitecture:
-					cm.Data[v1alpha1.PPCArchitecture] = v1alpha1.PPCArchitecture
-				}
+func (h *Handler) StoreCurrentValidConfig(condition v1alpha1.SamplesResourceConditionType, srcfg *v1alpha1.SamplesResource) error {
+
+	if condition != v1alpha1.SamplesExist {
+		return nil
+	}
+
+	cm, err := h.configmapclientwrapper.Get(h.namespace, v1alpha1.SamplesResourceName)
+	if err != nil && !kerrors.IsNotFound(err) {
+		// just return error to sdk for retry
+		return err
+	}
+	if kerrors.IsNotFound(err) {
+		// 4.0 testing showed that we were getting empty ConfigMaps
+		cm = nil
+	}
+	if cm != nil {
+		// just return ... we only update the config map once
+		return nil
+	}
+	cm = &corev1.ConfigMap{}
+	cm.Name = v1alpha1.SamplesResourceName
+	cm.Data = map[string]string{}
+	if len(srcfg.Spec.InstallType) == 0 {
+		cm.Data[installtypekey] = string(v1alpha1.CentosSamplesDistribution)
+	} else {
+		cm.Data[installtypekey] = string(srcfg.Spec.InstallType)
+	}
+	if len(srcfg.Spec.Architectures) == 0 {
+		cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
+	} else {
+		for _, arch := range srcfg.Spec.Architectures {
+			switch arch {
+			case v1alpha1.X86Architecture:
+				cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
+			case v1alpha1.PPCArchitecture:
+				cm.Data[v1alpha1.PPCArchitecture] = v1alpha1.PPCArchitecture
 			}
 		}
-		_, err = h.configmapclientwrapper.Create(h.namespace, cm)
-		if err != nil {
-			// just return error to sdk for retry
-			return err
-		}
 	}
-	return nil
+	_, err = h.configmapclientwrapper.Create(h.namespace, cm)
+	return err
 }
 
 func (h *Handler) SpecValidation(srcfg *v1alpha1.SamplesResource) error {
+	// TODO - the first thing this should do is check that all the config values
+	// are "valid" (the architecture name is known, the distribution name is known, etc)
+	// if that fails, we should immediately error out and set ConfigValid to false.
+
+	// only if the values being requested are valid, should we then proceed to check
+	// them against the previous values(if we've stored any previous values)
+
 	// if we have not had a valid SamplesResource processed, allow caller to try with
 	// the srcfg contents
-	if h.samplesResource == nil || !h.samplesResource.ConditionTrue(v1alpha1.SamplesExist) {
+	if !srcfg.ConditionTrue(v1alpha1.SamplesExist) {
+		logrus.Println("Spec is valid because samples are not created yet")
 		return nil
 	}
 	cm, err := h.configmapclientwrapper.Get(h.namespace, v1alpha1.SamplesResourceName)
 	if err != nil && !kerrors.IsNotFound(err) {
-		// just return error to sdk for retry
-		return err
+		err = fmt.Errorf("error retrieving previous configuration: %v", err)
+		return h.processError(srcfg, v1alpha1.ConfigurationValid, corev1.ConditionUnknown, err, "%v")
```

Comment:
we tried to validate the configuration, we failed to do so, the condition is now unknown (even if it was previously explicitly valid or invalid).

Reply:
It is the relist in particular I am concerned about wrt duplicate events.

And yes, the sdk.Update(...) and mutating / duplicate ... a bit more precisely, my experience was that I was *NOT* updating the spec, just the metadata for the keys, and the call to update to set the conditions was resulting in duplicate events ... so either
a) I missed a spot in the code, and my assertion is false ... in any event ,removing the "cached" version should help prove things out one way or the other
b) the sdk really needs an `UpdateStatus` like what I see in client-go, so it leaves the resource version alone

Reply:
this will be a potential add on to the follow up PR

Reply:
My understanding is that any change to the stored object will change the ResourceVersion value.  It's basically the etcd transaction identifier.  It has no understanding of whether you changed metadata, spec, or status.

So basically any time you update Status (ie setting a condition) you are going to get a new SampleResource event, with a new resourceversion, which your handler will need to process.  In general processing it should be a no-op (nothing has changed about the desired config), though since the Conditions changed, the ClusterOperator status update flow will likely be triggered to reflect the new Conditions into the ClusterOperator object.

Reply:
(so again, I don't think you were seeing "duplicate events", i think you were seeing the update event which *appeared* to contain the same resourceversion as the one you had cached, because when you called sdk.Update() your cached object's resource version got updated to the new resource version that resulted from the update.  Then the update event comes in, also with that resource version that's in your cached resource)